crates/supervisor/storage/src/providers/derivation_provider.rs

```diff
@@ -489,15 +489,97 @@ where
     /// This removes all derived blocks with number >= the given block number
     /// and updates the traversal state accordingly.
     pub(crate) fn rewind_to(&self, block: &BlockNumHash) -> Result<(), StorageError> {
+        info!(
+            target: "supervisor::storage",
+            chain_id = %self.chain_id,
+            target_block_number = %block.number,
+            target_block_hash = %block.hash,
+            "Starting rewind of derivation storage"
+        );
+
+        // Validate the block exists and get the block pair - this provides hash validation
         let block_pair = self.get_derived_block_pair(*block)?;
 
-        // Delete all derived blocks with number ≥ `block_info.number`
+        // Get the latest block number from DerivedBlocks
+        let latest_block = {
+            let mut cursor = self.tx.cursor_read::<DerivedBlocks>()?;
+            cursor.last()?.map(|(num, _)| num).unwrap_or(block.number)
+        };
+
+        // Check for future block
+        if block.number > latest_block {
+            error!(
+                target: "supervisor::storage",
+                chain_id = %self.chain_id,
+                target_block_number = %block.number,
+                latest_block,
+                "Cannot rewind to future block"
+            );
+            return Err(StorageError::FutureData);
+        }
+
+        let total_blocks = latest_block.saturating_sub(block.number);
+        const LOG_INTERVAL: u64 = 100;
```

Comment:
ah, so we also need to add a new field
```rust
pub(crate) struct DerivationProvider<'tx, TX> {
    tx: &'tx TX,
    chain_id: ChainId,
    #[doc(hidden)]
    observability_interval: usize,
}
```
so that in tests we can set this field very low. we will need an additional constructor `new_with_observability_interval` that takes an extra arg `observability_interval: Option<usize>`. in the existing constructor use this `LOG_INTERVAL` constant, which will have to be moved to the top of this module, and should be renamed to `DEFAULT_LOG_INTERVAL`.